pkg/registry/apis/provisioning/jobs/persistentstore.go

```diff
@@ -0,0 +1,466 @@
+	if err != nil {
+		return apifmt.Errorf("could not create requirement: %w", err)
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	jobsObj, err := s.jobStore.List(timeoutCtx, &internalversion.ListOptions{
+		LabelSelector: labels.NewSelector().Add(*requirement),
+		// We don't need to clean up everything all the time. Just do enough such that we have a fair amount of work.
+		Limit: 100,
+	})
+	cancel() // by the time we have the list, there is no response body to read, so just cancel immediately
+	if err != nil {
+		return apifmt.Errorf("failed to list jobs: %w", err)
+	}
+	jobs, ok := jobsObj.(*provisioning.JobList)
+	if !ok {
+		return apifmt.Errorf("unexpected object type %T", jobsObj)
+	}
+
+	for _, job := range jobs.Items {
+		if job.Labels == nil {
+			job.Labels = make(map[string]string)
+		}
+		delete(job.Labels, LabelJobClaim)
+		job.Status.State = provisioning.JobStatePending
+
+		// We list jobs from all namespaces. So when we want to update a specific job, we also need its namespace in the context.
+		ctx := request.WithNamespace(ctx, job.GetNamespace())
+		// Likewise, we should use the provisioning identity now that we have the namespace we are operating within.
+		ctx, _, err = identity.WithProvisioningIdentity(ctx, job.GetNamespace())
+		if err != nil {
+			// This should never happen, as it is already a valid namespace from the job existing... but better be safe.
+			return apifmt.Errorf("failed to get provisioning identity for '%s': %w", job.GetNamespace(), err)
+		}
+
+		timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		_, _, err := s.jobStore.Update(timeoutCtx,
+			job.GetName(),                       // name
+			rest.DefaultUpdatedObjectInfo(&job), // objInfo
+			failCreation,                        // createValidation
+			nil,                                 // updateValidation
+			false,                               // forceAllowCreate
+			&metav1.UpdateOptions{},             // options
+		)
+		cancel() // we have no response body to read, so just cancel immediately
+		if apierrors.IsConflict(err) || errors.Is(err, errWouldCreate) {
+			continue
+		}
+		if err != nil {
+			return apifmt.Errorf("failed to unclaim job '%s' in '%s': %w", job.GetName(), job.GetNamespace(), err)
+		}
+	}
+
+	return nil
+}
+
+// Insert adds a new job to the queue.
+//
+// The job name is not honoured. It will be overwritten with a name that fits the job.
+//
+// This saves it if it is a new job, or fails with `apierrors.IsAlreadyExists(err) == true` if one already exists.
+func (s *persistentStore) Insert(ctx context.Context, job *provisioning.Job) (*provisioning.Job, error) {
+	s.generateJobName(job) // Side-effect: updates the job's name.
+
+	obj, err := s.jobStore.Create(ctx, job, nil, &metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		return nil, apifmt.Errorf("job '%s' in '%s' already exists: %w", job.GetName(), job.GetNamespace(), err)
```

Comment:
I think we should catch this one on the callers to insert gracefully to say "already one in progress" (all good).

Reply:
agree. That's the intent of why I pass through the error :)